client/src/GameState.ts

```diff
@@ -64,31 +70,131 @@ export class GameState {
+            if (
+                this.adjacentWithPlayers(this.currentTetromino, [
+                    this.otherTetrominoes[i],
+                ])
+            ) {
+                this.emitAndPlaceCurrentTetromino();
+            }
         });
     }
 
+    public emitPlayerMove() {
+        this.socket.emit(
+            "playerMove",
+            this.playerId,
+            this.currentTetromino.reportState()
+        );
+    }
+
+    public emitAndPlaceCurrentTetromino() {
+        // place on board and emit events to the server
+        this.socket.emit(
+            "playerPlace",
+            this.playerId,
+            this.currentTetromino.reportState()
+        );
+        this.placeTetromino(this.currentTetromino);
+        // start a new tetromino from the top
+        this.currentTetromino.respawn();
+        // broadcast new tetromino position
+        this.emitPlayerMove();
+    }
+
     public placeTetromino(tetromino: Tetromino) {
-        tetromino.tiles.forEach((tile) => {
-            const [row, col] = [
-                tetromino.position[0] + tile[0],
-                tetromino.position[1] + tile[1],
-            ];
+        tetromino.tiles.forEach(([row, col]) => {
             this.board[row][col] = tetromino.type;
         });
     }
+
+    public moveIfCan(
+        movement: (tetro: Tetromino) => TetrominoLookahead
+    ): boolean {
+        // look-ahead for the next tetromino state after movement
+        const lookahead = movement(this.currentTetromino);
+
+        if (
+            this.overlapWithBoard(lookahead) ||
+            this.overlapWithPlayers(lookahead, this.otherTetrominoes)
+        ) {
+            return false;
+        }
+
+        this.currentTetromino.updateFromLookahead(lookahead);
+        return true;
+    }
+
+    /**
+     * check against static board, see if newTetro is overlapping with static monominoes placed on board
+     * @returns boolean - if `newTetro` overlaps with any blocks on the board
+     */
+    private overlapWithBoard(lookahead: TetrominoLookahead): boolean {
+        return lookahead.tiles.some(([row, col]) => {
+            return this.board[row][col] != null;
+        });
+    }
+
+    /**
+     * check against other players' tetrominoes: see if newTetro is overlapping with other existing players
+     * @returns boolean - if `newTetro` overlaps with any of the `tetrominoes`
+     */
+    private overlapWithPlayers(
+        lookahead: TetrominoLookahead,
+        tetrominoes: Array<Tetromino>
+    ): boolean {
+        return tetrominoes.some((tetro) => {
+            // if the bounding boxes are more than the-maximum-"manhattan distance" away
+            const isFarAway =
+                Math.abs(tetro.position[0] - lookahead.position[0]) +
+                    Math.abs(tetro.position[1] - lookahead.position[1]) >
+                8; // 8 is twice the max width of a tetromino bounding box
```

Comment:
Curious about the choice of Manhattan distance here... Couldn't the two axis be compared independently and verified that both are greater than 4? I think that would consider more cases such as when two horizontal `I` tetrominoes are side by side (below), allowing for more frequent shortcutting

```
[ - | - | - | - ] [ - | - | - | - ]
```

Reply:
I think you meant we check that on **any** of the axis the distance is greater than 4?